[feature] Barcode Scanner

ruialmeida51 opened the pull request:
Uses google api vision to create an easy way to scan all types of barcodes.

ruialmeida51 committed:
[feature] barcode scanner - uses google api vision to create an easy way to scan all types of barcodes.

pedro-belchior commented on barcodescanner/src/main/res/layout/fragment_barcode_reader.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Does this need to be a LinearLayout ? why not a FrameLayout ?

ruialmeida51 replied:
Not it does not need to, changed to a frame layout :)

pedro-belchior commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/inline/View.kt:
```diff
@@ -0,0 +1,15 @@
+        override fun onGlobalLayout() {
+            if (measuredWidth > 0 && measuredHeight > 0) {
+                viewTreeObserver.removeOnGlobalLayoutListener(this)
+                f()
```
Can this have a more specific name please ?

neteinstein replied:
Actually since this is an extension for a View, probably it should pass to the Kotlin module of Skeletoid

ruialmeida51 replied:
Moved to kt-extensions :)

pedro-belchior commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,333 @@
+    //Camera FPS
+    private var cameraFramesPerSeconds = 30.0f
+    //The amount of time the camera will detect barcodes after the first detection.
+    private var cameraDetectionDelay = 2000L
```
These default values should all be resources.

ruialmeida51 replied:
Done

ruialmeida51 committed:
Create README.md

neteinstein commented on barcodescanner/proguard-rules.pro:
```diff
@@ -0,0 +1,21 @@
+
+# If you keep the line number information, uncomment this to
+# hide the original source file name.
+#-renamesourcefileattribute SourceFile
```
Nothing needed?

ruialmeida51 replied:
Nope, all good here

neteinstein commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,333 @@
+
+            initCamera()
+        } else {
+            requestCameraPermission()
```
Add this as a default behavior but with the possibility of being overriden via a callback.
So if you add a callback for this, it will call it, and the callback will do whatever needed to deal with it, if not, call this.

ruialmeida51 replied:
For the whole block (init & request permissions) or just to request permissions?

neteinstein replied:
I'd say just the requestPermission, wdyt?

ruialmeida51 replied:
Yeah, I agree. Added the callback 👍

neteinstein commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,333 @@
+    }
+
+    override fun onResume() {
+        Log.w(BARCODE_FRAGMENT_TAG, "Re-opening the camera onResume.")
```
Please avoid using `Log`, always use `LOG` since it's Skeletoid default logger.

You already have Skeletoid base as a dependency, so it's just a simple change.

ruialmeida51 replied:
done!

neteinstein commented on barcodescanner/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,35 @@
+<resources>
+
+    <string name="barcode_scanner_fragment_reading_barcode_info">
```
We should be able to customize those from outside the module.

I'm not sure this is viable, if I can add those Ids to the app and it would compile, I don't think so. We should check a viable way to do this.

neteinstein replied:
Actually it's quite simple, please add to the docs
https://stackoverflow.com/questions/4263259/replace-or-override-string-in-android-library-project

neteinstein commented on barcodescanner/src/test/java/com/mindera/skeletoid/barcodescanner/ExampleUnitTest.java:
```diff
@@ -0,0 +1,17 @@
+package com.mindera.skeletoid.barcodescanner;
```
Delete this please.

neteinstein commented on barcodescanner/README.md:
```diff
@@ -0,0 +1,84 @@
+# Barcode Scanner
```
Missing reference to this module on the base README file.

bruno-garces commented on barcodescanner/README.md:
```diff
@@ -0,0 +1,84 @@
+
+## Using the barcode scanner
+
+1. Implement the `BarcodeReaderCallback` interface on your view. The `onBarcodeRead`callback will return the closest barcode to the center of the screen.
```
missing space before the word callback

bruno-garces commented on barcodescanner/README.md:
```diff
@@ -0,0 +1,84 @@
+      - Default: 1920
+      - Type: Int
+      
+    - `BARCODE_CAMERA_RESOLUTION_WIDTH_BUNDLE_KEY`Camera height resolution
```
missing space before the word Camera

bruno-garces commented on barcodescanner/README.md:
```diff
@@ -0,0 +1,84 @@
+            transaction.add(R.id.container, fragment, BarcodeReaderFragment.BARCODE_FRAGMENT_TAG)
+            transaction.addToBackStack(null)
+            transaction.commit()
+
```
no need for this empty space 👾

ruialmeida51 replied:
done 👍

bruno-garces commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/barcode/BarcodeGraphic.java:
```diff
@@ -0,0 +1,106 @@
+package com.mindera.skeletoid.barcodescanner.barcode;
+
+/*
+ * Copyright (C) The Android Open Source Project
```
if this in from the AOSP why do we need to copy the classes? Isn't there an API to import?

ruialmeida51 replied:
So we can extend some behaviours, such as drawing a rectangle around the scanned barcode, for example, on the BarcodeGraphic class.

bruno-garces commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/inline/View.kt:
```diff
@@ -0,0 +1,15 @@
+        override fun onGlobalLayout() {
+            if (measuredWidth > 0 && measuredHeight > 0) {
+                viewTreeObserver.removeOnGlobalLayoutListener(this)
+                f()
```
f is not a name is a letter

ruialmeida51 committed:
moved aftermeasured to kotlin ext and changed name to after drawn
added possibility of overriding request permissions behaviour
changed linear layout to framelayout
changed static values to res files
changed Log to LOG

ruialmeida51 committed:
Update README.md

neteinstein commented on barcodescanner/src/androidTest/java/com/mindera/skeletoid/barcodescanner/ExampleInstrumentedTest.java:
```diff
@@ -0,0 +1,26 @@
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+@RunWith(AndroidJUnit4.class)
+public class ExampleInstrumentedTest {
```
This is probably also useless. 

Maybe delete?

neteinstein commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/camera/CameraSource.java:
```diff
@@ -0,0 +1,1216 @@
+                    // quickly after stop).
+                    mProcessingThread.join();
+                } catch (InterruptedException e) {
+                    Log.d(TAG, "Frame processing thread interrupted on release.");
```
Please replace all `Log` with `LOG` also in these classes.

bruno-garces commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,333 @@
+        //Checks if the device has Google Play Services enabled.
+        val googleApiAvailability = GoogleApiAvailability.getInstance()
+        val isGoogleApiAvailable = googleApiAvailability.isGooglePlayServicesAvailable(activity?.applicationContext)
+        if (isGoogleApiAvailable != ConnectionResult.SUCCESS) googleApiAvailability.getErrorDialog(activity, isGoogleApiAvailable, RC_HANDLE_GOOGLE_MOBILE_SERVICES)
```
can we have brackets here?

ruialmeida51 replied:
with all gooste

bruno-garces commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,335 @@
+        if (::source.isInitialized) {
+            try {
+                sourcePreview.start(source, overlay)
+            } catch (throwable: Throwable) {
```
can't we know which exceptions are we trying to catch here?

bruno-garces commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,335 @@
+
+@Suppress("Unused")
+class BarcodeScannerFragment(private val barcodeCallback: BarcodeScannerCallback,
+                             private val permissionsCallback: BarcodeScannerPermissionsCallback?) : Fragment(), BarcodeGraphicTracker.BarcodeDetectorListener {
```
```
class BarcodeScannerFragment(
         private val barcodeCallback: BarcodeScannerCallback,
         private val permissionsCallback: BarcodeScannerPermissionsCallback?
) : Fragment(), BarcodeGraphicTracker.BarcodeDetectorListener {
```
Can we have it like this? ☝️

bruno-garces commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,335 @@
+     * Distance to center using pythagoras theorem.
+     *
+     * We want the hypotenuse which is c^2, so, c = sqrt(a^2 + b^2)
+     *
```
remove this extra line

ruialmeida51 replied:
OCD cured

bruno-garces commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,335 @@
+    /**
+     * Converts a barcode bounding box to a Rect object.
+     * a^2 + b^2 = c^2
+     *
```
remove this extra line

bruno-garces commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,335 @@
+        //To avoid that, RACE CONDITIONS!
+        synchronized(lock) {
+            //Avoid duplicates & null data..
+            if (data != null && readBarcodes.find { it.rawValue == data.rawValue } == null) {
```
`readBarcodes.none { it.rawValue == data.rawValue }`

pedro-belchior commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,335 @@
+    }
+
+    //Flags that represent the barcode formats we want to read.
+    private var barcodeFormats: Int = R.integer.barcodeFormats
```
Pretty sure you cant do this. You'll be using the ID int value instead of the resource value.

pedro-belchior commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,335 @@
+    //Camera width resolution
+    private var cameraResolutionWidth = R.integer.cameraResolutionWidth
+    //Camera height resolution
+    private var cameraResolutionHeight = R.integer.cameraResolutionHeight
```
Pretty sure you cant do this, this will fetch the int reference to the resource.

ruialmeida51 replied:
yup, fixed 👍

bruno-garces commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,335 @@
+    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
+        if (requestCode == RC_HANDLE_CAMERA_PERMISSION) {
+            if (grantResults.isNotEmpty() && grantResults.first() == PackageManager.PERMISSION_GRANTED) initCamera()
+            else Snackbar.make(snackbarView, R.string.barcode_scanner_fragment_no_camera_permission, Snackbar.LENGTH_SHORT).show()
```
if/else with brackets please

bruno-garces commented on barcodescanner/src/main/res/layout/fragment_barcode_reader.xml:
```diff
@@ -0,0 +1,22 @@
+                android:layout_width="match_parent"
+                android:layout_height="match_parent"/>
+
+
```
empty spaces 👾

bruno-garces commented on barcodescanner/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,35 @@
+    </string>
+
+    <string name="barcode_scanner_fragment_request_camera_permission_dont_ask_again">
+        You have chosen "Don't ask again". Please change your device settings to grant camera access to use this feature.
```
don't we need to escape this character?

bruno-garces commented on barcodescanner/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,35 @@
+    </string>
+
+    <string name="barcode_scanner_fragment_snackbar_confirm">
+        OK
```
can't we just use `@android:string/ok` ?

bruno-garces commented on kt-extensions/src/main/java/com/mindera/skeletoid/kt/extensions/views/View.kt:
```diff
@@ -3,10 +3,9 @@
-import android.view.View.GONE
-import android.view.View.INVISIBLE
-import android.view.View.VISIBLE
+import android.view.View.*
```
wildcard imports 🤢

bruno-garces commented on kt-extensions/src/main/java/com/mindera/skeletoid/kt/extensions/views/View.kt:
```diff
@@ -49,4 +48,15 @@ fun View.setPaddingBottom(paddingBottom: Int) {
+        override fun onGlobalLayout() {
+            if (measuredWidth > 0 && measuredHeight > 0) {
+                viewTreeObserver.removeOnGlobalLayoutListener(this)
+                f()
```
🙈

ruialmeida51 committed:
Update README.md

ruialmeida51 committed:
deleted useless tests
replaced Log with LOG on missing classes
fixed R.integer
improved naming
removed empty space to cure OCD

ruialmeida51 committed:
readme & typo

neteinstein commented on barcodescanner/src/main/java/com/mindera/skeletoid/barcodescanner/view/BarcodeScannerFragment.kt:
```diff
@@ -0,0 +1,335 @@
+    //Triggered when we read the first barcode. Used to start cameraDetectionDelay countdown.
+    private val hasBarcode = MutableLiveData<Boolean>() //Ensure this variable is only changed ONCE.
+    //To sync barcode detections and avoid dupes.
+    private val lock = Any()
```
Instead of this you can probably do the lock on BarcodeScannerFragment::class

neteinstein commented on README.md:
```diff
@@ -38,6 +38,9 @@ This library is **divided into modules** to enable you to just use just those th
 ### [App Rating module](https://github.com/Mindera/skeletoid/blob/master/apprating/)
 - Do you want to ask you users to rate to your app? Here you go, quick and easy!
 
+### [App Rating module](https://github.com/Mindera/skeletoid/blob/master/barcodescanner/)
```
App rating module? :-D

ruialmeida51 committed:
fixed readme
fixed ocd spaces
added b r a c k e t s
removed wildcard imports

pedro-belchior commented on barcodescanner/src/test/java/com/mindera/skeletoid/barcodescanner/ExampleUnitTest.kt:
```diff
@@ -0,0 +1,17 @@
+
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
```
Are we going to delete this ?

ruialmeida51 replied:
They are needed to pass the travis CI tests, in the future i'll do some proper testing
